### PR TITLE
The diagram respecting some name strategies are not included in the page containing the {{diagram /}} macro

### PR DESCRIPTION
**Pull request by Farcasut**
This PR aims to fix issues that arise when using different naming strategies. To achieve this, I've added a small function that will update the macro block to always contain a reference that respects the naming strategy.

To apply the current name strategy both of these options have to be enabled:

  * Validate names before saving
  *  Transform names automatically 
image
I Administration -> Editing -> Name Strategies

### CODE REVIEW

File: application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java [hunk 1]
```diff
@@ -65,26 +73,65 @@ public class DiagramMacroSolrMetadataExtractor implements SolrEntityMetadataExtr
...
+        try {
 
+            XWikiContext context = contextProvider.get();
+            boolean modfied = false;
+            for (Block macroBlock : macroBlocks) {
+                String referenceName =
+                    macroBlock.getParameter(REFERENCE) != null ? macroBlock.getParameter(REFERENCE) : "Diagram";
+                // For backwards compatibility we check if the page already exists so we won't modify it.
+                DocumentReference macroReference =
+                    explicitDocumentReferenceResolver.resolve(referenceName, document.getDocumentReference());
```
Comment (Farcasut) [inline application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java:110]: If we want to maintain backward compatibility, we have three cases:

    The reference does not follow the naming strategy, and the page is already created. In this case, we check if the page exists, and if it does, we retain the current reference.

    The reference does not follow the naming strategy, and the diagram was created in versions 1.20.7 or 1.20.8. In this case, the check for an existing page will fail, and the reference will be updated to follow the naming strategy. So instead of having {diagram reference="aaaa"/}, and transforming the naming strategy every time we load the macro, the macro call will become {diagram reference="bbbb"/} when reindexing the wiki, in the case of a naming strategy that transforms all "a"s into "b"s.

    The reference does not follow the naming strategy, and the page has not been created yet. In this case, when the wiki is reindexed, the reference will simply be updated to follow the naming strategy.

File: application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java [hunk 2]
```diff
@@ -99,4 +146,15 @@ private String transformName(String name)
             return name;
         }
     }
+
+    private boolean isValid(String name)
+    {
+        if (this.entityNameValidationConfiguration.useValidation()
```
Comment (Farcasut) [inline application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java:152]: If we want to transform the names even if the  "Validate names before saving" is disabled I can delete the check for this.entityNameValidationConfiguration.useValidation().

**Review by oanalavinia — CHANGES_REQUESTED** [review]

File: application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java [hunk 3]
```diff
@@ -65,26 +73,65 @@ public class DiagramMacroSolrMetadataExtractor implements SolrEntityMetadataExtr
...
+                    // Update the macro block to contain the valid reference.
+                    if (!isValid) {
+                        macroBlock.setParameter(REFERENCE, transformedNamed);
+                        modfied = true;
+                    }
+                }
+            }
+            if (modfied) {
+
+                document.setContent(xdom);
```
[the remark this answers is not in the document]
Comment (Farcasut) [reply application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java:127]: Yes, the List<BlockMacros> actually contains the original macroBlocks and the changes, directly affect the xdom.

File: application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java [hunk 4]
```diff
@@ -65,32 +73,81 @@ public class DiagramMacroSolrMetadataExtractor implements SolrEntityMetadataExtr
...
 
+            XWikiContext context = contextProvider.get();
+            boolean modified = false;
+            for (Block macroBlock : macroBlocks) {
+                String referenceName =
+                    macroBlock.getParameter(REFERENCE) != null ? macroBlock.getParameter(REFERENCE) : "Diagram";
+                // For backwards compatibility we check if the page already exists so we won't modify it.
+                DocumentReference macroReference =
+                    explicitDocumentReferenceResolver.resolve(referenceName, document.getDocumentReference());
+                boolean diagramAlreadyExists = context.getWiki().exists(macroReference, context);
```
Comment (oanalavinia) [inline application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java:120]: What happens if the document exists and the reference is not ok? We still display it and let the name handling to the diagram itself, right? A comment would be good

File: application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java [hunk 5]
```diff
@@ -65,32 +73,81 @@ public class DiagramMacroSolrMetadataExtractor implements SolrEntityMetadataExtr
...
+    }
 
+    /**
+     * Checks and updates the references of all the diagram macro calls to make sure that all of them respect the
+     * current name strategy.
+     *
+     * @param document document with all the macro calls
+     * @param xdom of the @document
+     * @param macroBlocks list of all the diagram macro calls
+     * @return true if all the references are valid, false otherwise.
```
Comment (oanalavinia) [inline application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java:106]: ` true if references are valid` sounds misleading. What I understand is that it returns true if references are not valid and they have been updated, false if there were no invalid references

File: application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java [hunk 6]
```diff
@@ -65,32 +73,81 @@ public class DiagramMacroSolrMetadataExtractor implements SolrEntityMetadataExtr
     @Inject
     private EntityNameValidationConfiguration entityNameValidationConfiguration;
 
+    @Inject
+    private Logger logger;
+
     @Override
     public boolean extract(XWikiDocument document, SolrInputDocument solrDocument)
```
Comment (oanalavinia) [inline application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java:80]: I realized you are missing unit tests for this class (+ what has been added in https://github.com/xwikisas/application-diagram/commit/1148b01f3ab5018a5c947cc39fc41ef83adf5414 )
Could you add in this PR unit tests for usecases specific to it, and make a separate PR for what other unit tests are missing?

File: application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java [hunk 7]
```diff
@@ -65,32 +73,81 @@ public class DiagramMacroSolrMetadataExtractor implements SolrEntityMetadataExtr
...
+                }
+            }
+            if (modified) {
+                document.setContent(xdom);
+                context.getWiki()
+                    .saveDocument(document, "Updated diagram macro references to respect the name strategy.", context);
+                return modified;
+            }
+        } catch (XWikiException e) {
+            logger.error(e.getMessage(), e);
```
Comment (oanalavinia) [inline application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java:140]:
```suggestion
            logger.error("Failed to update diagram macro references of [{}] to respect the naming strategy.", document, e)
```

File: application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java [hunk 8]
```diff
@@ -65,32 +73,81 @@ public class DiagramMacroSolrMetadataExtractor implements SolrEntityMetadataExtr
...
+            throw new RuntimeException(e);
```
Comment (oanalavinia) [inline application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java:141]: I'm not sure you need to throw an error. Maybe just `return false`?

File: application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java [hunk 9]
```diff
@@ -65,32 +73,81 @@ public class DiagramMacroSolrMetadataExtractor implements SolrEntityMetadataExtr
...
+                // For backwards compatibility we check if the page already exists so we won't modify it.
+                DocumentReference macroReference =
+                    explicitDocumentReferenceResolver.resolve(referenceName, document.getDocumentReference());
+                boolean diagramAlreadyExists = context.getWiki().exists(macroReference, context);
+                if (!diagramAlreadyExists) {
+                    // First we check if the name is valid in the current naming strategy.
+                    boolean isValid = this.isValid(referenceName);
+                    // If the name is valid then we can use it, otherwise we transform the name in a valid
+                    // one and update the macro block.
+                    if (!isValid) {
```
Comment (oanalavinia) [inline application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java:126]: I would add a `logger.debug` message for this case

**Review by oanalavinia — CHANGES_REQUESTED** [review]

File: application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java (hunk 4 above)
Comment (Farcasut) [reply application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java:120]: If the document exists and the reference doesn't respect the naming strategy the macro will be skipped and keep the current reference to breaking the macro calls that already exist.

File: application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java (hunk 5 above)
Comment (Farcasut) [reply application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java:106]: You are right the message should have been "true if any reference is invalid, false otherwise"

File: application-diagram-api/src/test/java/com/xwiki/diagram/internal/handlers/TestDiagramMacroSolrMetadataExtractor.java [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.when;
+
+/**
+ * Unit test for {@link DiagramMacroSolrMetadataExtractor}
+ */
+@ComponentTest
+public class TestDiagramMacroSolrMetadataExtractor
```
Comment (oanalavinia) [inline application-diagram-api/src/test/java/com/xwiki/diagram/internal/handlers/TestDiagramMacroSolrMetadataExtractor.java:56]: `DiagramMacroSolrMetadataExtractorTest`

File: application-diagram-api/src/test/java/com/xwiki/diagram/internal/handlers/TestDiagramMacroSolrMetadataExtractor.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+    @Mock
+    private SolrInputDocument solrDocument;
+
+    /**
+     * The diagram was created in another version, but the reference doesn't respect the name strategy, but we want to
+     * keep the reference so we don't break the macros that already exist.
+     */
+    @Test
+    public void updateReferenceCase1() throws XWikiException
```
Comment (oanalavinia) [inline application-diagram-api/src/test/java/com/xwiki/diagram/internal/handlers/TestDiagramMacroSolrMetadataExtractor.java:102]: No need to specify public here (for the other methods as well)

File: application-diagram-api/src/test/java/com/xwiki/diagram/internal/handlers/TestDiagramMacroSolrMetadataExtractor.java [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+            documentReference);
+        when(xwikiContext.getWiki().exists(documentReference, xwikiContext)).thenReturn(false);
+        when(entityNameValidationManager.getEntityReferenceNameStrategy().isValid("AAAA")).thenReturn(false);
+        when(entityNameValidationManager.getEntityReferenceNameStrategy().transform("AAAA")).thenReturn("BBBB");
+        extractor.extract(this.xwikiDocument, this.solrDocument);
+        assertEquals("BBBB", block.getParameter(REFERENCE));
+    }
+
+    @BeforeEach
+    void setup()
```
Comment (oanalavinia) [inline application-diagram-api/src/test/java/com/xwiki/diagram/internal/handlers/TestDiagramMacroSolrMetadataExtractor.java:153]: This could be added at the beginning of the class, since this is actually the first one called

File: application-diagram-api/src/test/java/com/xwiki/diagram/internal/handlers/TestDiagramMacroSolrMetadataExtractor.java [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        when(entityNameValidationManager.getEntityReferenceNameStrategy().transform("Strategy")).thenReturn(
+            "Strategy2");
+        extractor.extract(this.xwikiDocument, this.solrDocument);
+        assertEquals("Strategy", block.getParameter(REFERENCE));
+    }
+
+    /**
+     * The macro has the default settings where the reference is missing and there is a naming strategy that updates
+     * the name.
+     *
```
Comment (oanalavinia) [inline application-diagram-api/src/test/java/com/xwiki/diagram/internal/handlers/TestDiagramMacroSolrMetadataExtractor.java:119]: I think you forgot to apply one more time de checkstyle of your IDEA

File: application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java (hunk 5 above)
Comment (oanalavinia) [reply application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java]: I still find it incomplete. What about `true if any reference was invalid and has been updated, false if there weren't any invalid references or if an error occurred while updating the document`

File: application-diagram-api/src/test/java/com/xwiki/diagram/internal/handlers/TestDiagramMacroSolrMetadataExtractor.java [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+
+    @BeforeEach
+    void setup()
+    {
+        when(this.xwikiDocument.getXDOM()).thenReturn(xdom);
+        when(this.contextProvider.get()).thenReturn(this.xwikiContext);
+        when(this.xwikiContext.getWiki()).thenReturn(this.wiki);
+        when(this.xwikiDocument.getDocumentReference()).thenReturn(documentReference);
+        when(entityNameValidationConfiguration.useValidation()).thenReturn(true);
+        when(entityNameValidationConfiguration.useTransformation()).thenReturn(true);
```
Comment (oanalavinia) [inline application-diagram-api/src/test/java/com/xwiki/diagram/internal/handlers/TestDiagramMacroSolrMetadataExtractor.java:160]: I'll add a test for when these false

**Review by oanalavinia — CHANGES_REQUESTED** [review]

**Review by oanalavinia — APPROVED** [review]
